mkfs/rootdir.c

```diff
@@ -362,6 +369,75 @@ static int add_symbolic_link(struct btrfs_trans_handle *trans,
 	return ret;
 }
 
+static ssize_t zlib_compress_extent(struct btrfs_inode_item *btrfs_inode,
+				    u32 sectorsize, const void *in_buf,
+				    size_t in_size, void *out_buf)
+{
+	int ret;
+	z_stream strm;
+
+	strm.zalloc = Z_NULL;
+	strm.zfree = Z_NULL;
+	strm.opaque = Z_NULL;
+	ret = deflateInit(&strm, g_compression_level);
+	if (ret != Z_OK) {
+		error("deflateInit failed: %s", strm.msg);
+		return -EINVAL;
+	}
+
+	strm.next_out = out_buf;
+	strm.avail_out = BTRFS_MAX_COMPRESSED;
+	strm.next_in = (void *)in_buf;
+	strm.avail_in = in_size;
+
+	/* Try to compress the first sector - if it would be larger, mark the
+	 * inode as nocompress and return. */
+	if (!(btrfs_stack_inode_flags(btrfs_inode) & BTRFS_INODE_COMPRESS)) {
+		strm.avail_in = sectorsize;
+
+		ret = deflate(&strm, Z_SYNC_FLUSH);
+
+		if (ret != Z_OK) {
+			error("deflate failed: %s", strm.msg);
+			return -EINVAL;
+		}
+
+		if (strm.avail_out < BTRFS_MAX_COMPRESSED - sectorsize) {
+			u64 flags;
+
+			flags = btrfs_stack_inode_flags(btrfs_inode);
+			flags |= BTRFS_INODE_NOCOMPRESS;
```

Comment:
I prefer not to pass a btrfs inode and handle its flags inside a compression function.

Such compression function should only do the compression. And let the caller handling the btrfs specific things.
(E.g. if the compressed size is larger than the input,  return `-E2BIG` to inform the caller just like what kernel does).